[all] Fix the cause of failling test scenes (for a greener dashboard)

damienmarchal opened the pull request:
______________________________________________________
<!--- Please leave this at the end of your message -->
This PR: 
- [x] builds with SUCCESS for all platforms on the CI.
- [x] do generate LESS warnings.
- [x] do generate LESS unit test failures.
- [x] do generate LESS scene test failures.
- [ ] does not break API compatibility.
- [ ] is more than 1 week old (or has fast-merge label).

**Reviewers will merge only if all these checks are true.**

damienmarchal committed:
[SofaImplicitField] Remove files that shouldn't be there.

Those file have been merge by mistake some time ago but they have no
meaning in the current Sofa version. So I remove them.

damienmarchal committed:
[scripts/ci] Add "-DPLUGIN_SOFAEULERIANFLUID=ON" on the CI.

Otherwise some test scenes using Fluid2D & Fluid3D are failing.

damienmarchal committed:
[SofaDistanceGrid] Fix the FFDDistanceGridCollision.scn

The <include> directive was using local path instead of relative to the 'example'
directory as other examples.

NB: I wonder why the shared "Object" directory is in examples. If it is shared.
it should be in the 'share' directory ?

damienmarchal committed:
[Flexible] Remove from the test the BezierTetrahedron.scn as it depend on a plugin.HighOrder

NB: It would be better to generate the .scene-tests from the cmake
depending on the other plugin found.

damienmarchal committed:
[examples] Remove the <EDGESET> in RgulardGridTopology_dimensions.scn as this make the scene fail.

This scene is failing since its addition in
PR: https://github.com/sofa-framework/sofa/pull/270

damienmarchal committed:
[SofaKernel] Force printlog=true when verbose=true otherwise nothing is printed.

This is implemened only at init & reinit.

damienmarchal committed:
[example] Fix SparseGridRamificationTopology.scn that is failing since PR#521

damienmarchal committed:
[examples] Fix DeformableOnRigidFrameMappingConstraint.scn that fails since PR #521

The more I see PR the more I wonder why the algorithm allow 'den'
to be zero. Shouldn't this issue a warning saying that the scene is
probably broken ?

damienmarchal committed:
[plugins/MultiThreading] Fix initialization of mTaskMutex.v_

The plugin is broken since PR #549 as it removes the initialization
of the mutex. Possible explaination is that v_ is implementation
dependent but the existing initialization was done incorrectly
using a (int).

Now use BOOST_DETAIL_SPINLOCK_INIT which insure that the initialization
is coherent with the implementation used.

(I also cannot prevent myself to remove the two ugly commented lines)

damienmarchal committed:
[Flexible] Fix examples/material/StVenantKirchoff.scn broken since the merge of #PR 243

In #PR 243 the following change was introduced in MechanicalObject.inl::init():
```cpp
  if (!l_topology && d_useTopology.getValue())
    {
        l_topology.set( this->getContext()->getActiveMeshTopology() );
    }
```

In place of
```cpp
    //Look at a topology associated to this instance of MechanicalObject by a tag
    this->getContext()->get(m_topology, this->getTags());
    // If no topology found, no association, then look to the nearest one
    if(m_topology == NULL)
        m_topology = this->getContext()->getMeshTopology();
```

I assume that the in the scene was the EdgeSetTopologyContainer is not returned by
the getActiveMeshTopology method.

NB: All those implicit linking looks really problematic to me as they are error prone.

damienmarchal committed:
[SofaKernel] Remove compilation warning in MechanicalObject.inl + replace some sout with msg_info.

damienmarchal committed:
[WIP] CGLinearSolver ...le check sur zero est bizzare.

damienmarchal committed:
[MultiThreading] FIX TaskSchedulerBoost on macosX.

damienmarchal committed:
[MultiThreading] Another attempt to fix multithreading with boost 1.64 (WIP)

MacOs is hard to fix.
But fundamentally...using anything from boost::details is fundamentally
bad.

guparan commented on applications/plugins/MultiThreading/src/TaskSchedulerBoost.h:
```diff
@@ -165,7 +161,7 @@ namespace sofa
 		
 
-			mutable boost::detail::spinlock		mTaskMutex;
+            mutable boost::detail::spinlock		mTaskMutex {BOOST_DETAIL_SPINLOCK_INIT};
```
GG for the fix!
I think you should choose between keeping a TAB indentation or fixing it with spaces for the whole file ;-)

damienmarchal replied:
Well...I'm not in favor of changing the whole file indent for a one line modification. In that case isn't it better to apply the astyle just before merging the PR ? 

Shouldn't I wait for the automatic application of the astyle formatting sheet by the CI ?

damienmarchal commented on SofaKernel/modules/SofaBaseLinearSolver/CGLinearSolver.inl:
```diff
@@ -231,6 +251,20 @@ void CGLinearSolver<TMatrix,TVector>::solve(Matrix& M, Vector& x, Vector& b)
+         if( fabs(den) == 0.0 )
+         {
+             endcond = "div by zero. You scene is probably buggy.";
+             if( verbose )
```
Well...

yes but no but yes but no :) 

the only totally redundent thing is:
```cpp
if(printLog.getValue())
     msg_info() << "Blahblah"
```

printLog is the way to activate or deactive *all* the messages going through msg_info() but as it is implemented there is no fine grain control.  This is why adding an additional option like "verbose" (boolean or int) allows to implement multiple level of 'info' verbosity for your component. 

In sofa there is a lot of component that hard code multi-levels verbosity but in my component I prefer to stick to printLog as it make things easier and more consistant.

damienmarchal replied:
This PR was initially on fixing the failing test...thus I'm not sure it is good if we add too much cleaning or refactoring or cosmetic changes as it will make the review harder.  I just cannot prevent myself to force the printLog when verbose is set to simplify a bit the things (in init & reinit).

damienmarchal commented on SofaKernel/modules/SofaBaseLinearSolver/CGLinearSolver.inl:
```diff
@@ -231,6 +251,20 @@ void CGLinearSolver<TMatrix,TVector>::solve(Matrix& M, Vector& x, Vector& b)
+             if( verbose )
+             {
+                 sout<<"CGLinearSolver, den = "<<den<<", smallDenominatorThreshold = "<<f_smallDenominatorThreshold.getValue()<<sendl;
+             }
```
Of  course, if you think it is relevant you can fix it in another PR, just do it, but please keep in mind that if everyone open PR to fix scenes it will be a big mess very quickly and we will have to "ping pong" between all the PRs. 

EDIT: pushing regularly & working in this single PR have the advantage that we can see each other progress and activity. 

EDIT: I didn't responded clearly but yes you can fix everything you want I didn't planned to work on this one and the commited version was just there to show you the problems.

hugtalbot committed:
ADD RequiredPlugin for StVenantKirchhoff.scn

ErwanDouaille committed:
fix applications/plugins/OptiTrackNatNet/examples/optitrackGrasping.scn: [ERROR]   [RequiredPlugin]  failed to load: optiTrackNatNet

ErwanDouaille commented:
@damienmarchal I was looking for errors with this scene : 

> ./runSofa -g batch -s dag -n 100 /home/douaille/SOFA/fixTestScene/applications/plugins/Flexible/examples/demos/steak.scn

The segfault comes from : https://github.com/sofa-framework/sofa/commit/e65ef095fd96b69b3c9036b2e839f9429b4b022b#diff-179a1b721f69963e4d1fd5ee846ffe22R57

What does this line ? I tried remove it but exception_mode is used everywhere in Cimg.h and for its next call it crash.

damienmarchal committed:
[Flexible] FIX by adding ignore "shapeFunction/BezierTetrahedron.scn" in the CI as it depend on highOrderTopology.

damienmarchal commented:
@ErwanDouaille I look at the line/commit you pointed...but un don't understand why you are pointing to a specific commit that didn't changed that specific line.

ErwanDouaille commented:
@damienmarchal Because it crash at this line, no specific reason for commit sry, I was investigating ;)
I don´t understand what does this line do. And why does it crash there

damienmarchal commented:
Ok thank for the extra details. After a quick glance this line is changing the way CImg is handling error. 
The problem is that it tries to lock mutex which somehow cause the crash. 
My current guess is that if it could be related to the fact that there is either image & CImage plugin which are both using CImg...which have static initialized structure...just a guess...

EDIT: mmm I'm deeply thinking...in CImg
```cpp
#if defined(cimg_module)
    Mutex_info& Mutex_attr();
#elif defined(cimg_main)
    Mutex_info& Mutex_attr() { static Mutex_info val; return val; }
#else
    inline Mutex_info& Mutex_attr() { static Mutex_info val; return val; }
#endif
```

hugtalbot committed:
Revert "[WIP] CGLinearSolver ...le check sur zero est bizzare."

This will be fixed in PR 556

This reverts commit 46463887bc8919440b63c8f55d1a00232c11b3e7.

damienmarchal committed:
[FLEXIBLE] Use msg_info instead of serr in BaseDeformationMapping

serr is bad:
- it is inconsistent and emits a warning while it is named serr.
- it does not add the filename/line from where the message is emitted

In addition in this specific component the prints are looking like info/debug
message and I see ne reason why they should be error or warning.

damienmarchal committed:
[InvertibleFVM] Clean this plugin.

In this PR I:
- replaced serr by msg_error or msg_info (to avoid flooding the CI with useless warning message).
- remove commented code all around
- move the code from .h to .inl.
- add SOFA_TARGET and automatic component handling in initPlugin.cpp

hugtalbot committed:
Revert "[SofaKernel] Force printlog=true when verbose=true otherwise nothing is printed."

This reverts commit c04b7c28fc0ef234508add0519b60c61c62679b3
since the correction is already done in PR 556

hugtalbot committed:
Revert "[example] Fix SparseGridRamificationTopology.scn that is failing since PR#521"

This reverts commit 7d032c7b8155606da08990b2308e2d4292263554
since the scene will work thanks to fix in PR 556

damienmarchal committed:
[Flexible, image, CImg, CImgPlugin] Refactor to FIX crashes.

Problem:
CImg is a .h only library that contains a mutex to allow multi-threading
use of the library. The problem is that when the CImg/CImg.h is included
in several different shared library (i.e. plugin) then it is unclear
where the mutex are and when they are initialized. This scenario is visible
in Flexible where there is scenes (demos/steak.scn) that use both image
& CImgPlugin (as CImgPlugin is now a plugin & mandatory).

To fix that I refactor how CImg is included in the software.
The idea is to a have a single definition of CImg's mutex in
initCImgPlugin.cpp (so in CImgPlugin.so) and have all other
libraries that uses CImg to:
find_package(CImgPlugin REQUIRED) (instead of CImg)
then
target_link_libraries(${PROJECT_NAME} XX YYY ZZ CImgPlugin)

and in .h do:

[CImgPlugin] Clean a bit the whole CMakeLists.txt.

- CImgPlugin.h.in is useless (there is no value expanded)
- the test is now externatlized from the plugin (with correct dependencies to it)
- Remove useless header file inclusion.
- Remove commented code

[CimgPlugin] Centos WIP ?

[CImgPlugin] (WIP) Try to force the linking of the mutex for CImg.

[image] Trying to fix the link issues with CImgPlugin.

[Cimg] how to export properly the symbol on windows ?

[examples] Fix example that were using drawColorTetrahedra with three values instead of four.

[CImg] MOVE extlibs/CImg into extlibs/CImgPlugin

It is not allowed to used CImg directly anymore as having multiple
instances cause crashes...use SOFACImg instead.

[CImgPlugin/CImg] Update CMakeLists.txt to handle the move of CImg.

damienmarchal committed:
[BaseDeformationMapping] Only emit non empty message & replace sout with msg_info

msg_info & sout are equivalent (as sout is implemented on top of msg_info).
But msg_info have richier info.

damienmarchal committed:
[Compliant] Replace sout/serr with msg_info/msg in KrylovSolver and activate printing when verbose is true

the serr is generating warning which does not really make sense in this context.

damienmarchal committed:
[SofaPython] Fix a typo in fontain.py that generates warning in the CI.

damienmarchal committed:
[SofaGeneralDeformable] Refactor the printing of triangle flipped in TriangularBiquadraticSpringsForceField

guparan commented:
@damienmarchal, @fredroy: Why isn't CImg directly inside CImgPlugin? We would then just have to depend on CImgPlugin to use CImg.

damienmarchal commented:
In my fix I initially didn't wanted to change too much existing things in Sofa but as It is in fact a complex issue I think that guillaume suggestion to have CImg in CImgPlugin is a very good one and I will give it a try.

damienmarchal commented:
[ci-build][with-scene-tests]

damienmarchal committed:
[SofaConstraint] FIX compilations warning in XXXConstraintCorrection.

damienmarchal committed:
[SofaGeneralDeformable] Clean messages in TriangularBiquadraticSpringForceField

damienmarchal commented:
Dashboard is showing good progress, only one scene crash on ubuntu & centos in an easy to fix issue.
Still one unknown test-failure on centos (but it may be a configuration problem related to LANG).
One test failure on windows is still because this machine does not have libjpeg/libtiff installed (or we need a full build ?). Three are test problem

The big dark spot is now the 26 remaining crashes ? Can someone with MacOS investigate the issues please ?

untereiner commented:
A quick test of this branch on macos:  
- `./bin/runSofa -g batch -s dag -n 100 ../../sofa/examples/Demos/caduceus.scn` crashes with a segfault
- `./bin/runSofa -s dag -n 100 ../../sofa/examples/Demos/caduceus.scn` runs ok

There is call to `glGetIntegerv(GL_READ_FRAMEBUFFER_BINDING, &windowId);` who ends to a segfault.

damienmarchal commented:
Thanks for the feedback @untereiner maybe it is because in batch mode there is no opengl context.  
If I remember correctly this crash macos's tests that way. 

Can you providing line/file that cause problem ?

damienmarchal committed:
[SofaGeneralTopology] Add points to CubeTopology.

The points are now generated in the topology and can thus be used by the
mechanical model.

Note: I wonder how this was scene has ever worked. If any one knows please
tell, because my fix may not be the right way.

damienmarchal committed:
[SofaGraphComponent] Improve the SceneChecker to minimize the amount of warning generated.

When the scene is strongly problematic a lot of warning are issued.
In this new version we aggreate all the result into a single warning.

damienmarchal commented:
[ci-build][with-scene-tests]

damienmarchal committed:
[SceneChecker] FIX compilation issue related to improperly clearing a stringstream.

damienmarchal commented:
[ci-build][with-scene-tests]

guparan commented:
caduceus.scn crashes on both macos_default and macos_options on master when running in batch mode.  
See https://ci.inria.fr/sofa-ci/job/mac_clang-3.4_default/1904/warnings2Result/category.94921639/package.-1174366838/

damienmarchal commented:
@guparan Thank for the report.
As pointed by @untereiner  ...the problem is related to the batch mode & opengl function on macos.  
My guess is that there is not opengl context and that there is code in sofa that use opengl functions.

untereiner commented:
Below is the stack. I do not understand why the visual objects are created in the batch mode.
PS: do a search of "batch" in sofa... 

>  frame #0: 0x00007fff35526c44 libGL.dylib`glGetIntegerv + 18
 frame #1: 0x000000011ef3eac6 libSofaHelper_d.17.12.dev.dylib`sofa::helper::gl::FrameBufferObject::getCurrentFramebufferID() at FrameBufferObject.cpp:79
   frame #2: 0x00000001176c0882 libSofaOpenglVisual_d.17.12.dev.dylib`sofa::component::visualmodel::Light::Light(this=0x000000012080c800, vtt=0x0000000117a0a620) at Light.cpp:86
  frame #3: 0x00000001176c6b7b libSofaOpenglVisual_d.17.12.dev.dylib`sofa::component::visualmodel::PositionalLight::PositionalLight(this=0x000000012080c800, vtt=0x0000000117a0a618) at Light.cpp:633
    frame #4: 0x00000001176c84cf libSofaOpenglVisual_d.17.12.dev.dylib`sofa::component::visualmodel::SpotLight::SpotLight(this=0x000000012080c800) at Light.cpp:698
    frame #5: 0x00000001176d7ae9 libSofaOpenglVisual_d.17.12.dev.dylib`sofa::core::objectmodel::New<sofa::component::visualmodel::SpotLight>::New<>(this=0x00007ffeefbf59c8) at SPtr.h:57
    frame #6: 0x00000001176d79e5 libSofaOpenglVisual_d.17.12.dev.dylib`sofa::core::objectmodel::New<sofa::component::visualmodel::SpotLight>::New<>(this=0x00007ffeefbf59c8) at SPtr.h:57
    frame #7: 0x00000001176d789f libSofaOpenglVisual_d.17.12.dev.dylib`sofa::component::visualmodel::SpotLight::SPtr sofa::core::objectmodel::BaseObject::create<sofa::component::visualmodel::SpotLight>((null)=0x0000000000000000, context=0x0000000122011a00, arg=0x0000000121261df0) at BaseObject.h:100
    frame #8: 0x00000001176d7763 libSofaOpenglVisual_d.17.12.dev.dylib`sofa::core::ObjectCreator<sofa::component::visualmodel::SpotLight>::createInstance(this=0x000000012071ba90, context=0x0000000122011a00, arg=0x0000000121261df0) at ObjectFactory.h:209
    frame #9: 0x000000011bb62cb8 libSofaCore_d.17.12.dev.dylib`sofa::core::ObjectFactory::createObject(this=0x000000011c7be9f0, context=0x0000000122011a00, arg=0x0000000121261df0) at ObjectFactory.cpp:186
    frame #10: 0x000000011aca2c54 libSofaSimulationCommon_d.17.12.dev.dylib`sofa::core::ObjectFactory::CreateObject(context=0x0000000122011a00, arg=0x0000000121261df0) at ObjectFactory.h:157
    frame #11: 0x000000011aca0a65 libSofaSimulationCommon_d.17.12.dev.dylib`sofa::simulation::xml::ObjectElement::initNode(this=0x0000000121261df0) at ObjectElement.cpp:77
    frame #12: 0x000000011aca0177 libSofaSimulationCommon_d.17.12.dev.dylib`sofa::simulation::xml::ObjectElement::init(this=0x0000000121261df0) at ObjectElement.cpp:60
    frame #13: 0x000000011ac8b407 libSofaSimulationCommon_d.17.12.dev.dylib`sofa::simulation::xml::BaseElement::init(this=0x000000012125f6e0) at BaseElement.cpp:149
    frame #14: 0x000000011ac9de25 libSofaSimulationCommon_d.17.12.dev.dylib`sofa::simulation::xml::NodeElement::init(this=0x000000012125f6e0) at NodeElement.cpp:78
    frame #15: 0x000000011ac75d15 libSofaSimulationCommon_d.17.12.dev.dylib`sofa::simulation::SceneLoaderXML::processXML(xml=0x000000012125f6e0, filename="../../sofa/examples/Demos/caduceus.scn") at SceneLoaderXML.cpp:117
    frame #16: 0x000000011ac75479 libSofaSimulationCommon_d.17.12.dev.dylib`sofa::simulation::SceneLoaderXML::load(this=0x00000001212063e0, filename="../../sofa/examples/Demos/caduceus.scn") at SceneLoaderXML.cpp:79
    frame #17: 0x000000011aeafa79 libSofaSimulationCore_d.dylib`sofa::simulation::Simulation::load(this=0x000000012200d800, filename="../../sofa/examples/Demos/caduceus.scn") at Simulation.cpp:470
    frame #18: 0x00000001000213fc runSofa_d`main(argc=8, argv=0x00007ffeefbff928) at Main.cpp:381
    frame #19: 0x00007fff536e1115 libdyld.dylib`start + 1
    frame #20: 0x00007fff536e1115 libdyld.dylib`start + 1

damienmarchal commented:
@untereiner I think that this is because batch mode does not "remove" element from the scene 
(it would be a nighmare). To me the underlying problem is that the batch mode should have a valid opengl  context created (off-screen rendering) as we use it to test file with opengl based components. 

Thanks for the stack trace... I will try to make something about it (even on macos).

untereiner commented:
Launching an opengl context "by hand" is complicated if I remember correctly. I suggest that the batch mode uses GLEW to handle the opengl context. In the other options it is Qt that handles the context creation.
But I do not understand why it works on linux.

damienmarchal committed:
[SofaGraphComponent] CHANGE in the way RequiredPlugin handles inputs.

The old version of the plugin is loading as plugins the content
of pluginName or, if empty, the "name" of the component.

To improve functionnal consistency this, when the name is use as pluginName,
adds it to the pluginName lists. So that pluginName contains all the
"informations".

damienmarchal committed:
[SofaGraphComponent] Minor cleaning in SceneChecker

damienmarchal commented:
It is definitively hard to have an opengl context by hand...and having one for offscreen (without a windows) seems even more tricky (anyone's help is welcome). I know @ErwanDouaille search for that in his HeadLessRecorder

I assume it work on linux/windows...because they just don't crash/segfault when gl function are called without a buffer :) 
As said in:
https://www.opengl.org/discussion_boards/showthread.php/158904-OpenGL-function-calls-without-available-contexts

Too bad... I tried a lot of approach based on QOffscreenSurface and friends, they work on my system but as soon as I move that to a CI machine (which does not have X or GLX or whatever)...the application crash. I reseted the branch but you can still see the garbage in the Dashboard.  I will try a different approach tomorrow.

damienmarchal commented:
[ci-build][with-scene-tests]

damienmarchal commented:
Ho damn...two full build + scene that are green builds... and not one to cheers (deep sadness)
https://www.sofa-framework.org/dash/?branch=pr/fixTestScene

damienmarchal commented:
[ci-build][with-scene-tests]

guparan commented:
Wow, such green, very stable, many thanks! ;-)

damienmarchal commented:
More seriously...I don't know how to fix the MacOS version (and I think the batch mode is fundamentally broken :)). 

Tomorrow I will clean the PR, possibly cutting it in pieces because I doubt guys here will accept to merge it "as it is" :)

guparan commented:
@epernod Could you quick review the commit 47ef2e63a80c please? How could CubeTopology.scn have been passing scene-tests without this fix?
(request from @damienmarchal)

damienmarchal committed:
Merge remote-tracking branch 'upstream/master' into fixTestScene

guparan committed:
Merge remote-tracking branch 'sofa-framework/master' into fixTestScene

# Conflicts:
#	applications/plugins/CMakeLists.txt

guparan commented:
This PR seems quite empty now. Shall we close it @damienmarchal ?

damienmarchal commented:
Yes, 
Thank for the help in managing all that.

guparan commented:
Thank you for your hard work! :-)
